### PR TITLE
Transaction precommitwrite

### PR DESCRIPTION
**Pull request by triump2020**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #4693 

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/txn/storage/tae/storage.go [hunk 1]
```diff
@@ -26,71 +34,169 @@ import (
...
+		return nil, err
+	}
+	for _, entry := range precommitW.EntryList {
+		//err = doHandleCmd(txnHandle, entry)
+		err = txnHandle.HandleCmd(entry)
+	}
+	return nil, err
+}
+
+//func doHandleCmd(txn txnif.AsyncTxn, cmd *api.Entry) (err error) {
```
Comment (zhangxu19830126) [inline pkg/txn/storage/tae/storage.go:182]: The commented code needs to be removed directly

**Review by zhangxu19830126 — CHANGES_REQUESTED** [review]

File: pkg/txn/storage/tae/storage.go (hunk 1 above)
Comment (triump2020) [reply pkg/txn/storage/tae/storage.go:182]: ok

File: pkg/container/types/types.go [hunk 2]
```diff
@@ -116,6 +117,7 @@ type Uuid [16]byte
 // timestamp for transaction: physical time (higher 8 bytes) + logical (lower 4 bytes)
```
[the remark this answers is not in the document]
Comment (triump2020) [reply pkg/container/types/types.go:117]: OK

File: pkg/container/types/types.go [hunk 3]
```diff
@@ -79,8 +79,9 @@ const (
 )
 
 const (
-	TxnTsSize = 12
-	RowidSize = 16
+	TxnTsSize  = 14
+	TxnTidSize = 32
```
Comment (fengttt) [inline pkg/container/types/types.go:83]: I think you need at least update container/types/txnts.go -- that is where the TS is implemented.   Unfortunately it has some magic contants, that won't work with simply update TxnTsSize.

Comment (fengttt) [reply pkg/container/types/types.go:83]: Do you want to make Tid a type?   Like Rowid is [RowidSize]byte, make this a [TxnTidSize]byte, and update correct places.

File: pkg/txn/storage/tae/storage.go [hunk 4]
```diff
@@ -26,71 +35,144 @@ import (
...
+	tae       *db.DB
+	txns      struct {
+		sync.Mutex
+		idMap map[string]txnif.AsyncTxn
+	}
 }
 
-func New(
+func openTAE(targetDir string) (*db.DB, error) {
+	mask := syscall.Umask(0)
```
Comment (fengttt) [inline pkg/txn/storage/tae/storage.go:51]: This essentially won't work -- you have to use FileService API.    Or you need to make a sweeping change later.
At least, TAEStorage may be S3.

File: pkg/txn/storage/tae/storage.go [hunk 5]
```diff
@@ -26,71 +35,144 @@ import (
...
+		return err
+	}
+	return txnHandle.Commit()
+
 }
 
-// Destroy implements storage.TxnStorage
-func (*Storage) Destroy(ctx context.Context) error {
+// Committing implements storage.TxnTAEStorage
+func (s *TAEStorage) Committing(ctx context.Context, txnMeta txn.TxnMeta) error {
```
Comment (fengttt) [inline pkg/txn/storage/tae/storage.go:128]: This is fine -- but we prefer people to use panic(moerr.NewNYI("what is not implemented"))

File: pkg/vm/engine/tae/txn/txnimpl/txn.go [hunk 6]
```diff
@@ -63,3 +67,97 @@ func (txn *txnImpl) DatabaseNames() (names []string) {
...
+	db, err := txn.GetDatabase(cmd.DatabaseName)
+	if err != nil {
+		return err
+	}
+	//Handle DDL
+	if cmd.DatabaseId == catalog.SystemDBID {
+		switch cmd.TableId {
+		case catalog.SystemTable_DB_ID:
+			if cmd.EntryType == api.Entry_Insert {
+				//TODO:: build schema of database.
```
Comment (fengttt) [inline pkg/vm/engine/tae/txn/txnimpl/txn.go:108]: Comment out here are very bad idea -- you silently corrupt a database, right?   Missing insert ...

If you really need to not implement this time and leave it to next PR, panic here.

File: pkg/txn/storage/tae/storage.go (hunk 5 above)
Comment (triump2020) [reply pkg/txn/storage/tae/storage.go:128]: ok

File: pkg/txn/storage/tae/storage.go (hunk 4 above)
Comment (triump2020) [reply pkg/txn/storage/tae/storage.go:51]: Here is just for testing.

File: pkg/container/types/types.go (hunk 3 above)
Comment (triump2020) [reply pkg/container/types/types.go:83]: will  remove the Tid type,  use TxnMeta.ID which is a byte slice .

File: pkg/vm/engine/tae/txn/txnimpl/txn.go (hunk 6 above)
Comment (triump2020) [reply pkg/vm/engine/tae/txn/txnimpl/txn.go:108]: Ok,  i 'll use panic instead.

File: pkg/txn/storage/tae/log.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
+// Copyright 2022 Matrix Origin
```
Comment (fengttt) [inline pkg/txn/storage/tae/log.go:1]: You are not going to introduce another log, use logutil.   Talk to XZX if you need help.

File: pkg/txn/storage/tae/read.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+func (s *taeStorage) Read(
+	ctx context.Context,
+	txnMeta txn.TxnMeta,
+	op uint32,
+	payload []byte) (res storage.ReadResult, err error) {
+
+	switch op {
+
+	//case ...
```
Comment (fengttt) [inline pkg/txn/storage/tae/read.go:37]: I don't know what this // case comment gonna help you.   If it is really not implemented yet, panic in default.

File: pkg/txn/storage/tae/read.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+	var req Req
+	if err := gob.NewDecoder(bytes.NewReader(payload)).Decode(&req); err != nil {
+		return nil, err
+	}
+
+	var resp Resp
+	defer logReq("read", req, txnMeta, &resp, &err)()
+	defer func() {
+		if closer, ok := (any)(resp).(io.Closer); ok {
```
Comment (fengttt) [inline pkg/txn/storage/tae/read.go:73]: I will just put logutil ... stuff here, in one defer.

File: pkg/vm/engine/tae/containers/utils.go [hunk 10]
```diff
@@ -122,3 +126,201 @@ func CopyToMoVectors(vecs []Vector) []*movec.Vector {
...
+	taeV := MOToTAEVector(moV, nullable)
+	return taeV
+}
+
+func MOToTAEVector(v *movec.Vector, nullable bool) Vector {
+	vec := MakeVector(v.Typ, nullable)
+	bs := NewBytes()
+	switch v.Typ.Oid {
+	case types.T_bool:
+		if v.Col == nil || len(v.Col.([]bool)) == 0 {
```
Comment (fengttt) [inline pkg/vm/engine/tae/containers/utils.go:144]: You should not directly access v.Col, must MustTCol

**Review by fengttt — COMMENTED** [review]
I wonder if you can try to get rid of TAE vector.

WHY do we need so many of that?   Also, why do we need a proto vector?   Isn't vector serialized and deserialized to bytes -- that is all?

File: pkg/txn/storage/tae/log.go (hunk 7 above)
Comment (triump2020) [reply pkg/txn/storage/tae/log.go:1]: ok

File: pkg/txn/storage/tae/read.go (hunk 8 above)
Comment (triump2020) [reply pkg/txn/storage/tae/read.go:37]: OK

File: pkg/container/batch/tools.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+// limitations under the License.
+
+package batch
+
+import (
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/pb/api"
+)
+
+func BatchToProtoBatch(bat *Batch) (*api.Batch, error) {
```
Comment (zhangxu19830126) [inline pkg/container/batch/tools.go:22]: What is the difference between two batch and why can't we use one batch definition?

**Review by zhangxu19830126 — CHANGES_REQUESTED** [review]

File: pkg/vm/engine/tae/containers/utils.go (hunk 10 above)
Comment (triump2020) [reply pkg/vm/engine/tae/containers/utils.go:144]: ok

File: pkg/txn/storage/tae/read.go (hunk 9 above)
Comment (triump2020) [reply pkg/txn/storage/tae/read.go:73]: > I wonder if you can try to get rid of TAE vector.
> 
> WHY do we need so many of that? Also, why do we need a proto vector? Isn't vector serialized and deserialized to bytes -- that is all?

`Proto vector` had been  involved in  before this PR .

File: pkg/container/batch/tools.go (hunk 11 above)
Comment (triump2020) [reply pkg/container/batch/tools.go:22]: > I wonder if you can try to get rid of TAE vector.
> 
> WHY do we need so many of that? Also, why do we need a proto vector? Isn't vector serialized and deserialized to bytes -- that is all?

`Proto vector` had been involved in before this PR .

**Review by zhangxu19830126 — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Review by yingfeng — APPROVED** [review]

**Comment by triump2020** [comment]
@matrix-meow rerun failed

**Review by XuPeng-SH — APPROVED** [review]
Need some follow-up pr continuous improvement

**Comment by triump2020** [comment]
@matrix-meow rerun failed

**Pull request merged**